libcxx/test/std/algorithms/alg.nonmodifying/alg.find.last/ranges.find_last.pass.cpp

```diff
@@ -0,0 +1,294 @@
+                      if constexpr (std::forward_iterator<Iter>)
+                        test_iterators<Iter>();
+                      test_iterators<Iter, sentinel_wrapper<Iter>>();
+                      test_iterators<Iter, sized_sentinel<Iter>>();
+                    });
+                  });
+
+  {
+    std::vector<std::vector<int>> vec = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
+    auto view                         = vec | std::views::join;
+    assert(std::ranges::find_last(view.begin(), view.end(), 4).begin() == std::next(view.begin(), 3));
+    assert(std::ranges::find_last(view, 4).begin() == std::next(view.begin(), 3));
+  }
+
+  { // check that the last element is returned
+    {
+      struct S {
+        int comp;
+        int other;
+      };
+      S a[]    = {{0, 0}, {0, 2}, {0, 1}};
+      auto ret = std::ranges::find_last(a, 0, &S::comp).begin();
+      assert(ret == a + 2);
+      assert(ret->comp == 0);
+      assert(ret->other == 1);
+    }
+    {
+      struct S {
+        int comp;
+        int other;
+      };
+      S a[]    = {{0, 0}, {0, 2}, {0, 1}};
+      auto ret = std::ranges::find_last(a, a + 3, 0, &S::comp).begin();
+      assert(ret == a + 2);
+      assert(ret->comp == 0);
+      assert(ret->other == 1);
+    }
+  }
+
+  {
+    // check that an iterator is returned with a borrowing range
+    int a[]                                            = {1, 2, 3, 4};
+    std::same_as<std::ranges::subrange<int*>> auto ret = std::ranges::find_last(std::views::all(a), 1);
+    assert(ret.begin() == a);
+    assert(*ret.begin() == 1);
+  }
+
+  {
+    // count invocations of the projection
+    {
+      int a[]              = {1, 2, 3, 4};
+      int projection_count = 0;
+      auto ret             = std::ranges::find_last(a, a + 4, 2, [&](int i) {
+                   ++projection_count;
+                   return i;
+                 }).begin();
+      assert(ret == a + 1);
+      assert(*ret == 2);
+      assert(projection_count == 3);
+    }
+    {
+      int a[]              = {1, 2, 3, 4};
+      int projection_count = 0;
+      auto ret             = std::ranges::find_last(a, 2, [&](int i) {
+                   ++projection_count;
+                   return i;
+                 }).begin();
+      assert(ret == a + 1);
+      assert(*ret == 2);
+      assert(projection_count == 3);
+    }
+  }
+
+  return true;
+}
+
+template <class IndexT>
+class Comparable {
+  IndexT index_;
+
+public:
+  Comparable(IndexT i)
+      : index_([&]() {
+          IndexT size = static_cast<IndexT>(comparable_data.size());
+          comparable_data.push_back(i);
+          return size;
+        }()) {}
+
+  bool operator==(const Comparable& other) const { return comparable_data[other.index_] == comparable_data[index_]; }
+
+  friend bool operator==(const Comparable& lhs, long long rhs) { return comparable_data[lhs.index_] == rhs; }
+};
+
+void test_deque() {
```

Comment:
Can we parameterise this and test all the sequence containers, please?

Reply:
Yeah, I just copied these tests from `ranges::find`

Reply:
Actually, I'm not sure it's useful to test all the sequence containers - it seems mostly like wasted compute. How about just testing a forward, bidirectional, and random access container?

Reply:
@EricWF once made an argument that testing that this works with all of our containers is good for ensuring all of our containers can use all of our algorithms without issue (I think the idea is that different algorithms will expose different usage patterns and potentially expose latent bugs). It's not integration-level, but it does achieve something more than generating heat.

Reply:
I mean, I will do it if others agree with you here, but it just seems very unhelpful to me - this algorithm doesn't do anything interesting, really. All it is is `operator--`, `operator++`, and `operator*`